### PR TITLE
PortfolioPlanningDataService - runDependencyQuery

### PR DESCRIPTION
**Pull request by ezambra**

### CODE REVIEW

File: src/PortfolioPlanning/Common/Services/PortfolioPlanningDataService.ts [hunk 1]
```diff
@@ -550,6 +479,52 @@ export class PortfolioPlanningDataService {
...
+                return {
+                    exceptionMessage: errorMessage,
+                    Links: [],
+                    QueryInput: queryInput
+                };
+            }
+
+            return {
+                exceptionMessage: null,
+                Links: this.WorkItemLInksQueryResultItems(rawResponseValue.responseValue),
```
Comment (ezambra) [inline src/PortfolioPlanning/Common/Services/PortfolioPlanningDataService.ts]: `WorkItemLinksQueryResultItems`

File: src/PortfolioPlanning/Common/Services/PortfolioPlanningDataService.ts [hunk 2]
```diff
@@ -1058,4 +1035,260 @@ export class ODataQueryBuilder {
...
+                        const props = {
+                            ["WorkItemId"]: workItemId,
+                            ["ProjectId"]: projectIdKey
+                        };
+                        const actionName =
+                            "PortfolioPlanningDataService/runDependencyQuery/MissingWorkItemRollUpValues";
+                        console.log(`${actionName}. ${JSON.stringify(props, null, "    ")}`);
+                        PortfolioTelemetry.getInstance().TrackAction(actionName, props);
+                    } else {
+                        //  Check if this is a work item type we would like to show as dependecy.
```
Comment (ezambra) [inline src/PortfolioPlanning/Common/Services/PortfolioPlanningDataService.ts]: `dependency`

File: src/PortfolioPlanning/Common/Services/PortfolioPlanningDataService.ts [hunk 3]
```diff
@@ -1058,4 +1035,260 @@ export class ODataQueryBuilder {
 
         return result;
     }
+
+    /**
+     * 
+    $select=WorkItemLinkSK,SourceWorkItemId,TargetWorkItemId,LinkTypeReferenceName,ProjectSK
```
Comment (shutchings) [inline src/PortfolioPlanning/Common/Services/PortfolioPlanningDataService.ts]: Will we be using this?

File: src/PortfolioPlanning/Components/Plan/DependencyPanel.tsx [hunk 4]
```diff
@@ -56,15 +57,19 @@ export class DependencyPanel extends React.Component<IDependencyPanelProps, IDep
...
             dependencies => {
                 //Sort dependencies by target date
-                dependencies.DependsOn.sort((a, b) => (a.TargetDate > b.TargetDate) ? 1 : -1);
-                dependencies.HasDependency.sort((a, b) => (a.TargetDate > b.TargetDate) ? 1 : -1);
+                const projectIdKey = this.props.projectInfo.id.toLowerCase();
 
-                this._getWorkItemIcons(dependencies.DependsOn.concat(dependencies.HasDependency)).then(
+                let { DependsOn, HasDependency } = dependencies.byProject[projectIdKey];
+
+                DependsOn.sort((a, b) => (a.TargetDate > b.TargetDate ? 1 : -1));
```
[the remark this answers is not in the document]
Comment (shutchings) [reply src/PortfolioPlanning/Components/Plan/DependencyPanel.tsx]: Should we let the consumers sort? Or always sort them in the service?

File: src/PortfolioPlanning/Models/PortfolioPlanningQueryModels.ts [hunk 5]
```diff
@@ -182,10 +183,51 @@ export interface PortfolioPlanningWorkItemTypeFieldNameQueryResult extends IQuer
...
+    [projectIdKey: string]: { workItemIds: number[]; projectConfiguration: IProjectConfiguration };
 }
 
 export interface PortfolioPlanningDependencyQueryResult extends IQueryResultError {
-    DependsOn: PortfolioPlanningQueryResultItem[];
-    HasDependency: PortfolioPlanningQueryResultItem[];
+    byProject: {
+        [projectId: string]: {
+            //  successor
+            DependsOn: PortfolioPlanningQueryResultItem[];
```
[the remark this answers is not in the document]
Comment (ezambra) [reply src/PortfolioPlanning/Models/PortfolioPlanningQueryModels.ts]: ok, will wait on Xin and update.

[the remark this answers is not in the document]
Comment (ezambra) [reply src/PortfolioPlanning/Models/PortfolioPlanningQueryModels.ts]: never mind, Xin will merge after my changes.

File: src/PortfolioPlanning/Common/Services/PortfolioPlanningDataService.ts (hunk 3 above)
Comment (ezambra) [reply src/PortfolioPlanning/Common/Services/PortfolioPlanningDataService.ts]: yes, we are... I leave a comment on all these methods to remember what query I was trying to build :)

File: src/PortfolioPlanning/Components/Plan/DependencyPanel.tsx [hunk 6]
```diff
@@ -56,15 +57,19 @@ export class DependencyPanel extends React.Component<IDependencyPanelProps, IDep
...
+                HasDependency.sort((a, b) => (a.TargetDate > b.TargetDate ? 1 : -1));
+
+                this._getWorkItemIcons(DependsOn.concat(HasDependency)).then(
```
[the remark this answers is not in the document]
Comment (ezambra) [reply src/PortfolioPlanning/Components/Plan/DependencyPanel.tsx]: `ProjectConfiguration` should have icon information for all work item types, including dependencies as well. We are already retrieving icons when calling `ProjectConfigurationDataService.getProjectConfiguration` @LiXin2014  FYI

[the remark this answers is not in the document]
Comment (shutchings) [reply src/PortfolioPlanning/Components/Plan/DependencyPanel.tsx]: Nice, so we can delete the extra icon fetch code.

[the remark this answers is not in the document]
Comment (ezambra) [reply src/PortfolioPlanning/Components/Plan/DependencyPanel.tsx]: yeah, I am working on that change right now, will include in this PR

File: src/PortfolioPlanning/Components/Plan/DependencyPanel.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (ezambra) [reply src/PortfolioPlanning/Components/Plan/DependencyPanel.tsx]: pushed down sorting to the service

File: src/PortfolioPlanning/Components/Plan/DependencyPanel.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (ezambra) [reply src/PortfolioPlanning/Components/Plan/DependencyPanel.tsx]: done.

**Pull request merged**